ui build: split admin ui build based on license

benesch opened the pull request:
This is a ready-to-merge version of #19850 so that @mrtracy has a place to put his cluster visualization work. The approach is unchanged from @couchand's earlier work; I merely updated it for our new Makefile layout. I also moved the license type indicator to the debug page, as I think it's useful to have around until we have a more complete CCL feature set.

----

This change plumbs the OSS/CCL build choice through to the admin UI
build, to enable development of upcoming UI features covered by the CCL.
CCL code is expected to be encapsulated by hook components, each of
which will have a no-op version present in the OSS build.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20051)
<!-- Reviewable:end -->

mrtracy commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 12 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20051#-:-Kyw2nYIK5kYtogClogb:bmxilld)*
<!-- Sent from Reviewable.io -->

couchand commented:
I have to reiterate my concern that this doesn't check the license key.  Maybe that's okay for now, but it seems like it might be enough of an issue to be a blocker.

mrtracy commented:
The license key check is not a blocker, as long as its clearly understood that this code is *subject* to a license key check.

couchand commented:
OK, fair enough.  I still don't think we've got a clear idea of how we're going to resolve my [last question on the other PR](https://github.com/cockroachdb/cockroach/pull/19850#issuecomment-344021327) (and I'm likely not going to have the mental space to try to resolve those concerns this week), so I'm personally hesitant to merge a half-completed solution.

mrtracy commented:
If it helps to clarify the situation, not having this is blocking me from checking in a large body of code.

As for your last question, it's only an issue if OSS thingy has a direct counterpart in the ccl directory shadowing it.

```
/ccl/src
  multiplex.tsx
  fancyComponent.tsx
  badLicenceComponent.tsx
/src
  multiplex.tsx
  notFancyComponent.tsx

-- ccl/src/multiplex.tsx
if (!license.exists())
    return <notFancyComponent>
else if (!license.valid())
    return <BadLicenseComponent>
else
    return <fancyComponent>

-- src/multiplex.tsx
return <notFancyComponent>
```

You can still reference `notFancyComponent` from CCL because it's not shadowed.

couchand commented:
That's much uglier than originally conceived, but it does seem like it should work.  I'd personally like to see it implemented here in the example license check component before merging this, if that's okay.

I appreciate your concern about getting code merged.  If all that's required is a place to put code that should be covered by the CCL, we should just make a directory with nothing else in it except the relevant license file and not import it from anywhere yet.

benesch commented:
I'm a bit worried about the churn on the build system if we need to revise the approach, so in light of Andrew's concerns I think we should hold off on this until tomorrow.

benesch commented:
FWIW, @couchand, this PR does have a "license check": https://github.com/cockroachdb/cockroach/pull/20051/files#diff-74d9ac468667a0c83489332a0224623bR18

But all it does it look for a magic string in the URL.

mrtracy commented:
I don't consider this strategy "ugly"; it actually seems ideal in many ways.  You could just delete the `/ccl` directory and get an OSS build.

Referencing a shadowed OSS component is a non-issue to me; if you need to reference an OSS component from CCL code, don't shadow it.  However, it appears that Nikhil has a solution of adding "/oss" to the import path in order to explicitly request the OSS version of a component.

And to reiterate, the license check does *not* need to happen for this to be merged, we can enforce the check later with no issue.  What we *can't* casually do later is move code from OSS to CCL, so I need this sooner rather than later.

couchand commented:
Either of those solutions works.  This now has a working implementation, we can add the actual license check later, so LGTM.

benesch committed:
ui: split build based on license

This change plumbs the OSS/CCL build choice through to the admin UI
build to enable development of upcoming UI features covered by the CCL.
CCL code is expected to be encapsulated by hook components, each of
which will have a no-op version present in the OSS build.

As a side-effect of this change, the UI package now ships a stub UI with
no JavaScript dependencies. (The stub UI merely prints "Binary built
without web UI.") This means we can provide a new build target,
`buildshort`, which skips building the UI, and that `make test` no
longer needs to build the UI. 🎉

benesch commented:
I wasn't quite satisfied with the reviewed version of the code, so I hammered on the Go side of this solution some more, and I'm much happier with the result. The Webpack side of the solution is virtually unchanged.

In short, rather than generating `embedded_ccl.go` and `embedded_oss.go` in package `ui`, we generate two UI packages, `ui/distccl` and `ui/distoss`. Package`ui` instead exposes a stub UI that says "Binary built without web UI." which is overridden by init hooks in either `ui/distccl` or `ui/distoss`. This avoids everyone's concerns about build tags, as we choose which package to import by choosing an entrypoint (main.go vs cmd/cockroach-oss/main.go) instead of specifying build tags.

As an unexpected side effect, we no longer need to build the UI to run Go tests, which should make the non-UI developers here much happier!

couchand commented on Makefile:
```diff
@@ -523,36 +525,44 @@ ui-lint: | $(UI_PROTOS)
+# were both out-of-date. [0]
+#
+# XXX: Ideally we'd scope the dependency on $(UI_PROTOS) to the protos DLL, but
+# Make v3.81 has a bug that causes the dependency to be ignored [0]. We're stuck
```
`[1]`?

benesch replied:
Whoops! Thanks.

couchand commented on pkg/ui/process-bindata.sed:
```diff
@@ -0,0 +1,21 @@
+// GENERATED FILE DO NOT EDIT
+
+# Add pkg/ui import statement in import block.
+/^import (/a\
```
Could we just put this content in both `distccl/distccl.go` and `distoss/distoss.go`?

benesch replied:
Man, I wish. It would sure be cleaner, wouldn't it? Unfortunately because of the stupid way `go test pkg/...` works, it's essential that `pkg/ui/distccl` and `pkg/ui/distoss` are compilable whether or not the UI has been generated.

I'm somewhat comforted by the fact that one day the bug in goimports will be fixed, and the Makefile can just say:

```shell
echo 'func init() { ui.Asset = Asset; ui.AssetDir = AssetDir... }' > $@
goimports -s -w $@
```

couchand commented on pkg/ui/ui.go:
```diff
@@ -13,4 +13,57 @@
+
+// AssetInfo loads and returns metadata for the asset with the given name. It
+// returns an error if the asset could not be found or could not be loaded.
+var AssetInfo func(name string) (os.FileInfo, error)
```
Hmm... we should probably either not include this at all or write a stub implementation.  Leaving it nil seems unwise.

benesch replied:
It's apparently entirely legal for it to be nil. https://github.com/elazarl/go-bindata-assetfs/blob/30f82fa23fd844bd5bb1e5f216db87fd77b5eb43/assetfs.go#L149-L152

couchand commented on pkg/cmd/cockroach-short/main.go:
```diff
@@ -0,0 +1,23 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
```
I wonder if it's worth commenting what this is?

benesch replied:
Definitely. Done. (Or it will be in a moment.)

couchand reviewed (approved):
What magic you've wrought, sir!  You're about to get very popular.

benesch commented:
Ok, @mrtracy stamped this via Slack, so in it goes! I've tested this rev pretty extensively—building from a clean slate, building with intermediate results from an old branch, building with intermediate results from this branch—and it all seems to work. Please feel send additional complaints/feedback my way, though, and I can address them in future PRs.
